src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -116,6 +116,66 @@ void M(object? o, object o2)
                 );
         }
 
+        [Fact, WorkItem(29964, "https://github.com/dotnet/roslyn/issues/29964")]
+        public void IndexerUpdatedBasedOnReceiver_ReturnType_Inferred()
+        {
+            var comp = CreateCompilation(@"
+public class C<T>
+{
+    public T Field = default!;
+
+    void M(object? o, object o2)
+    {
+        if (o is null) return;
+        L(o)[0].Field.ToString();
+
+        o2 = null;
+        L(o2)[0].Field.ToString(); // 1
+    }
+
+    C<T> this[T index] { get => throw null!; set => throw null!; }
+
+    C<U> L<U>(U u) => null!;
+}", options: WithNonNullTypesTrue());
```

Comment:
Minor point: It's a little confusing reading when these tests have instance methods where the receiver is unused. For instance, I believe this could be written as:
```C#
class C<T>
{
    public T Field = default!;
    public C<T> this[T index] { get => throw null!; set => throw null!; }
}
class Program
{
    static void M(object? o, object o2)
    {
        //...
    }
    static C<U> L<U>(U u) => null!;
}
```